x-pack/plugin/vector-tile/src/main/java/org/elasticsearch/xpack/vectortile/rest/GridAggregation.java

```diff
@@ -0,0 +1,241 @@
+        }
+
+        @Override
+        public Rectangle toRectangle(String bucketKey) {
+            return GeoTileUtils.toBoundingBox(bucketKey);
+        }
+    },
+    GEOHEX {
+
+        // Because hex bins do not fit perfectly on a tile, we need to buffer our queries in order to collect
+        // all points inside the bin. For levels 0 and 1 we will consider all data, values for level 2 have
+        // been computed manually and approximated. The amount that the buffer decreases by level has been
+        // approximated to 2.5 (brute force computation suggest ~2.6 in the first 9 levels so 2.5 should be safe).
+        private static final double[] LAT_BUFFER_SIZE = new double[16];
+        private static final double[] LON_BUFFER_SIZE = new double[16];
+        static {
+            LAT_BUFFER_SIZE[0] = LAT_BUFFER_SIZE[1] = Double.NaN;
+            LON_BUFFER_SIZE[0] = LON_BUFFER_SIZE[1] = Double.NaN;
+            LAT_BUFFER_SIZE[2] = 3.7;
+            LON_BUFFER_SIZE[2] = 51.2;
+            for (int i = 3; i < LON_BUFFER_SIZE.length; i++) {
+                LAT_BUFFER_SIZE[i] = LAT_BUFFER_SIZE[i - 1] / 2.5;
+                LON_BUFFER_SIZE[i] = LON_BUFFER_SIZE[i - 1] / 2.5;
+            }
+        }
+        // Mapping between a vector tile zoom and a H3 resolution. The mapping tries to keep the density of hexes similar
+        // to the density of tile bins but trying not to be bigger.
+        // Level unique tiles H3 resolution unique hexes ratio
+        // 1 4 0 122 30.5
+        // 2 16 0 122 7.625
+        // 3 64 1 842 13.15625
+        // 4 256 1 842 3.2890625
+        // 5 1024 2 5882 5.744140625
+        // 6 4096 2 5882 1.436035156
+        // 7 16384 3 41162 2.512329102
+        // 8 65536 3 41162 0.6280822754
+        // 9 262144 4 288122 1.099098206
+        // 10 1048576 4 288122 0.2747745514
+        // 11 4194304 5 2016842 0.4808526039
+        // 12 16777216 6 14117882 0.8414913416
+        // 13 67108864 6 14117882 0.2103728354
+        // 14 268435456 7 98825162 0.3681524172
+        // 15 1073741824 8 691776122 0.644266719
+        // 16 4294967296 8 691776122 0.1610666797
+        // 17 17179869184 9 4842432842 0.2818666889
+        // 18 68719476736 10 33897029882 0.4932667053
+        // 19 274877906944 11 237279209162 0.8632167343
+        // 20 1099511627776 11 237279209162 0.2158041836
+        // 21 4398046511104 12 1660954464122 0.3776573213
+        // 22 17592186044416 13 11626681248842 0.6609003122
+        // 23 70368744177664 13 11626681248842 0.165225078
+        // 24 281474976710656 14 81386768741882 0.2891438866
+        // 25 1125899906842620 15 569707381193162 0.5060018015
+        // 26 4503599627370500 15 569707381193162 0.1265004504
+        // 27 18014398509482000 15 569707381193162 0.03162511259
+        // 28 72057594037927900 15 569707381193162 0.007906278149
+        // 29 288230376151712000 15 569707381193162 0.001976569537
```

Comment:
Since H3 only goes down to 1m² and clearly elastic square tiles go down to much smaller tiles, how feasible would it be to increase the number of H3 levels to match the resolution of the tiles? Perhaps two more H3 levels would suffice? Or is it that in H3 the 0..15 range is very clearly hard-coded (eg. only 4 bits used to store the H3 level)?

Reply:
I don't think it is just hard coded, e.g if you increase the max resolution all the test fail so there is much more into it. Note that hexagons don't divide the sphere perfectly so you have some pentagons (and other shapes) that need to fail the gaps.